### ISSUE TITLE
Circuit Breaking

### ISSUE THREAD

**Issue by ignasi35** [akka/akka-grpc#138]
Akka-gRPC client should allow setting up circuit breaking. Related to #137

**Comment by ignasi35** [comment]
Some mental notes:

 A gRPC client has a `Channel` instance connecting the client to a single remote host and that `Channel` is shared across several calls.

Lagom's `CircuitBreakingServiceLocator` offers a different circuit breaker instance for each call on a remote service. Different approaches could be: (1) per remote service or (2) per remote instance. Maybe akka-grpc could consider supporting all 3 alternatives in the future.


Akka provides general purpose [circuit breakers](https://doc.akka.io/docs/akka/2.5/common/circuitbreaker.html) so users can manually setup circuit breaking around their `GrpcService` calls if they want to.

Lagom provides a general purpose [circuit breaker panel](https://github.com/lagom/lagom/issues/1248) and _also_ promotes `ServiceLocator` implementors to extend [`CircuitBreakingServiceLocator`](https://www.lagomframework.com/documentation/1.4.x/scala/api/com/lightbend/lagom/scaladsl/client/CircuitBreakingServiceLocator.html) which keeps circuit breaking a bit hidden but still exposed to users via Settings. `CircuitBreakingServiceLocator` looks up a service instance and then runs userland code wrapped on a circuit breaker.

**Comment by yahor-filipchyk** [comment]
gRPC project has this approved [proposal](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) for retries that includes retry throttling. I think it'd be nice if `akka-grpc` client could use the configuration and semantics written in the proposal, opening the door for future full support of gPRC standard, including service config discovery via the DNS or other service discovery mechanisms.

**Comment by raboof** [comment]
I agree this is an area where the Akka gRPC client can be improved! We should definitely look at the linked proposal, but also remember to look for the best ways to translate this into Akka ecosystem concepts (such as akka-discovery and eventually akka-http)

**Comment by ignasi35** [comment]
After digging a bit more on the topic, I've found [a discussion](https://github.com/grpc/grpc-java/issues/7281) hinting the grpc-java client can support circuit breaking via `Interceptor`s (the discussion links to a [PoC in Kotlin using resilience4J](https://gist.github.com/eungju/226274b3dacb3203de3514bcf1c54505#file-circuitbreakerclientinterceptor-kt)).

Using `Interceptor`s is more convenient than my suggestion above because it sets up the circuit breaker per gRCP client automatically (instead of having to wrap every call on the client).

**Comment by ignasi35** [comment]
The implementation based on an `Interceptor` is not competing with the [proposal](https://github.com/grpc/proposal/blob/master/A6-client-retries.md?rgh-link-date=2019-09-04T16%3A17%3A40Z) but an alternative available in simpler deployments (where not DNS or similar config delivery mechanisms exist).

**Comment by ignasi35** [comment]
Hmm, maybe another possible approach would be to [wrap the method invocations in the generated code](https://github.com/akka/akka-grpc/blob/eb2906fba10c83ac0eba1bc61a2eb8276db50507/codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt#L89)?

**Comment by exp0nge** [comment]
Hey, any updates on this?

**Comment by He-Pin** [comment]
> Hey, any updates on this?

Pr is welcome😃😃😃
